### PR TITLE
mobile: compose payment links to sms/email

### PR DESCRIPTION
**Pull request by nalinbhardwaj**
Tested on iOS and Android, only works on real devices: https://streamable.com/wlj0zb

TODO:
- [x] rename Recipient interface to contact or friend or something else now that it multiplexes - done in #622

### CODE REVIEW

File: apps/daimo-mobile/src/view/shared/composeSend.tsx [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  recipientName?: string;
+};
+
+// TODO: In future, add invite link compose params
+type ComposeParams = PaymentLinkComposeParams;
+
+type ComposeSend = (sendParams: ComposeParams) => Promise<boolean>;
+
+const encodingFunction = Platform.OS === "ios" ? encodeURI : encodeURIComponent;
+const smsDivider = Platform.OS === "ios" ? "&" : "?";
```
Comment (nalinbhardwaj) [inline apps/daimo-mobile/src/view/shared/composeSend.tsx:18]: took a painful amount of time to get this right lol

File: apps/daimo-mobile/src/sync/recipients.ts [hunk 2]
```diff
@@ -1,26 +1,65 @@
...
+export interface AccountRecipient extends EAccount {
+  type: "account";
   originalMatch?: string;
   lastSendTime?: number;
 }
 
+interface EmailRecipient {
+  type: "email";
+  email: EmailAddress;
+  name?: string; // TODO: populate from contact book
```
Comment (dcposch) [inline apps/daimo-mobile/src/sync/recipients.ts:26]: we can do this by adding to `SystemContact`/`MessageContact` in #622

File: apps/daimo-mobile/src/view/shared/shareSheetURL.tsx [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import { Alert, Platform, Share, ShareAction } from "react-native";
+
+export async function shareSheetURL(
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/shared/shareSheetURL.tsx:3]: recommend eg

```ts
// Share via system share sheet = user picks messaging app
export async function shareURL(url: string) {
```
...comment, verb name, & the dollars argument looks unused

File: packages/daimo-common/src/model.ts [hunk 4]
```diff
@@ -82,3 +82,16 @@ export const zRecommendedExchange = z.object({
...
 
 export type RecommendedExchange = z.infer<typeof zRecommendedExchange>;
+
+export const zEmailAddress = z.string().email();
+
+export type EmailAddress = z.infer<typeof zEmailAddress>;
+
+// From https://stackoverflow.com/a/29767609
+const phoneNumberRegex = new RegExp(
+  /^\+?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4,6}$/im
```
Comment (dcposch) [inline packages/daimo-common/src/model.ts:92]: 👍

File: packages/daimo-common/src/model.ts [hunk 5]
```diff
@@ -82,3 +82,16 @@ export const zRecommendedExchange = z.object({
...
+);
+
+export const zPhoneNumber = z.string().regex(phoneNumberRegex);
+
+export type PhoneNumber = z.infer<typeof zPhoneNumber>;
```
Comment (dcposch) [inline packages/daimo-common/src/model.ts:97]: we may have to normalize phone numbers = strip everything but `+` and 0-9

File: apps/daimo-mobile/src/view/screen/send/SendNoteScreen.tsx [hunk 6]
```diff
@@ -51,19 +56,65 @@ export function SendNoteScreen({ route }: Props) {
...
+
+  const [externalAction, setExternalAction] = useState<ExternalAction>({
+    type: "share",
+    exec: shareSheetURL,
+  });
+
+  useEffect(() => {
+    if (!recipient) return;
+
+    const testAndGetComposer = async (): Promise<ExternalAction> => {
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/send/SendNoteScreen.tsx:70]: i think you can pull this into its own function below the componet

```
async function getSendLinkAction(recipient: Contact): Promise<ExternalAction> {
 // ...
}
```

less logic inside component always nice

File: apps/daimo-mobile/src/view/screen/send/RecipientDisplay.tsx [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import { View, StyleSheet } from "react-native";
+
+import { Recipient, getRecipientName } from "../../../sync/recipients";
+import { AccountBubble } from "../../shared/AccountBubble";
+import { ButtonCircle } from "../../shared/ButtonCircle";
+import Spacer from "../../shared/Spacer";
+import { useNav } from "../../shared/nav";
+import { TextH3, TextLight } from "../../shared/text";
+
+export function RecipientDisplay({
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/send/RecipientDisplay.tsx:11]: 👍

File: apps/daimo-mobile/src/view/screen/send/SendNoteScreen.tsx [hunk 8]
```diff
@@ -51,19 +56,65 @@ export function SendNoteScreen({ route }: Props) {
...
           title="Pay by sending a link"
-          subtitle="Anyone with the link can accept it"
+          subtitle="Daimo invite included with money"
         />
         <Spacer h={32} />
-        <TextCenter>
-          <TextLight>Enter amount</TextLight>
-        </TextCenter>
+        {recipient ? (
+          <RecipientDisplay recipient={recipient} />
```
Comment (dcposch) [inline apps/daimo-mobile/src/view/screen/send/SendNoteScreen.tsx:110]: 🚀

**Review by dcposch — APPROVED** [review]
LGreatTM after comments

File: packages/daimo-common/src/model.ts (hunk 5 above)
Comment (nalinbhardwaj) [reply packages/daimo-common/src/model.ts:97]: Yeah I looked it up before what the best recommendation is, it seems most people recommend against normalizing due to region variations: https://stackoverflow.com/a/41946

In the current way it leaves it to the system to normalize, which seems reasonable.

File: apps/daimo-mobile/src/view/shared/shareSheetURL.tsx (hunk 3 above)
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/view/shared/shareSheetURL.tsx]: done, thanks

File: apps/daimo-mobile/src/view/screen/send/SendNoteScreen.tsx (hunk 6 above)
Comment (nalinbhardwaj) [reply apps/daimo-mobile/src/view/screen/send/SendNoteScreen.tsx]: done

**Pull request merged**